octree/include/pcl/octree/octree_iterator.h

```diff
@@ -524,6 +523,77 @@ namespace pcl
       };
 
     //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    /** \brief @b Octree iterator class
+     * \note This class implements a forward iterator for traversing octrees in a breadth-first manner.
+     *       It only returns the node at a given depth in the octree, if any.
+     * \ingroup octree
+     * \author Fabien Rozar (fabien.rozar@gmail.com)
+     */
+    template<typename OctreeT>
+      class OctreeBreadthFirstAtDepthIterator : public OctreeBreadthFirstIterator<OctreeT>
+      {
+      public:
+
+        // public typedefs
+        typedef typename OctreeBreadthFirstIterator<OctreeT>::BranchNode BranchNode;
+        typedef typename OctreeBreadthFirstIterator<OctreeT>::LeafNode LeafNode;
+
+        /** \brief Empty constructor.
+          * \param[in] depth_arg Depth level during traversal
+          */
+        explicit
+        OctreeBreadthFirstAtDepthIterator (unsigned int depth_arg = 0);
+
+        /** \brief Constructor.
+        * \param[in] octree_arg Octree to be iterated. Initially the iterator is set to its root node.
+        * \param[in] depth_arg Depth level during traversal
+        */
+        explicit
+        OctreeBreadthFirstAtDepthIterator (OctreeT* octree_arg, unsigned int depth_arg = 0);
+
+        /** \brief Empty deconstructor. */
+        virtual
+        ~OctreeBreadthFirstAtDepthIterator ();
+
+        /** \brief Copy operator.
+        * \param[in] src the iterator to copy into this
+        */
+        inline OctreeBreadthFirstAtDepthIterator&
+        operator = (const OctreeBreadthFirstAtDepthIterator& src)
+        {
+          OctreeBreadthFirstIterator<OctreeT>::operator=(src);
+        }
+
+        /** \brief Reset the iterator to the first node at the depth given initially of the octree
+        */
+        void
+        reset ();
```

Comment:
I think it's useful to allow specifying the intended depth on reset. If you don't specify any argument it just resets to the first node of the current set depth, but if you do it switches the iterator to the first node of the new depth value.

Reply:
I make a try and I let you check it.